cli/command/stack/list.go

```diff
@@ -18,18 +22,92 @@ func newListCommand(dockerCli command.Cli) *cobra.Command {
 		Short:   "List stacks",
 		Args:    cli.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if dockerCli.ClientInfo().HasKubernetes() {
-				kli, err := kubernetes.WrapCli(dockerCli, kubernetes.NewOptions(cmd.Flags()))
-				if err != nil {
-					return err
-				}
-				return kubernetes.RunList(kli, opts)
-			}
-			return swarm.RunList(dockerCli, opts)
+			return runList(cmd, dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
 	flags.StringVar(&opts.Format, "format", "", "Pretty-print stacks using a Go template")
+	flags.StringSliceVar(&opts.Namespaces, "namespace", []string{"default"}, "Kubernetes namespaces to use")
+	flags.SetAnnotation("namespace", "kubernetes", nil)
+	flags.SetAnnotation("namespace", "experimentalCLI", nil)
+	flags.BoolVarP(&opts.AllNamespaces, "all-namespaces", "", false, "List stacks among all Kubernetes namespaces")
+	flags.SetAnnotation("all-namespaces", "kubernetes", nil)
 	return cmd
 }
+
+func runList(cmd *cobra.Command, dockerCli command.Cli, opts options.List) error {
+	var err error
+	stacks := []*formatter.Stack{}
+	if dockerCli.ClientInfo().HasSwarm() {
+		ss, err := swarm.GetStacks(dockerCli)
+		if err != nil {
+			return err
+		}
+		stacks = append(stacks, ss...)
+	}
+	if dockerCli.ClientInfo().HasKubernetes() {
+		if dockerCli.ClientInfo().HasAll() && !cmd.Flags().Changed("namespace") {
+			opts.AllNamespaces = true
+		}
+		if opts.AllNamespaces {
+			stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(cmd.Flags()))
+			if err != nil {
+				return err
+			}
+		} else {
+			mnms, err := getNamespaces(cmd)
+			if err != nil {
+				return err
+			}
+			for nm := range mnms {
+				stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(cmd.Flags(), nm))
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	var format string
+	switch {
+	case opts.Format != "":
+		format = opts.Format
+	case dockerCli.ClientInfo().HasKubernetes():
+		format = formatter.KubernetesStackTableFormat
+	default:
+		format = formatter.SwarmStackTableFormat
+	}
+	stackCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.NewStackFormat(format),
+	}
+	sort.Slice(stacks,
+		func(i, j int) bool {
+			return sortorder.NaturalLess(stacks[i].Name, stacks[j].Name)
+		})
+	return formatter.StackWrite(stackCtx, stacks)
+}
+
+func getStacks(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, kopts kubernetes.Options) ([]*formatter.Stack, error) {
```

Comment:
`getStacks` should take `stacks` as argument — instead, the caller should do the `append(…`, this would be far more readable.

Reply:
OK